packages/cactus-plugin-ledger-connector-polkadot/README.md

```diff
@@ -0,0 +1,200 @@
+
+const apiHost = `http://${address}:${port}`;
+
+const apiConfig = new Configuration({ basePath: apiHost });
+
+const apiClient = new PolkadotApi(apiConfig);
+
+const req: RunTransactionRequest = {
+  // See tests for specific examples on request properties
+};
+
+try {
+  const res = await apiClient.runTransaction(req);
+} catch (ex: Error) {
+  // Make sure to handle errors gracefully (which is dependent on your use-case)
+  console.error(ex);
+  throw ex;
+}
+```
+## 2. Architecture
+The sequence diagrams for various endpoints are mentioned below
+
+### 2.1. run-transaction-endpoint
+
+![run-transaction-endpoint sequence diagram](docs/architecture/images/run-transaction-endpoint.png)  
+The above diagram shows the sequence diagram of run-transaction-endpoint. User A (One of the many Users) interacts with the API Client which in turn, calls the API server. API server then executes transact() method which is explained in detailed in the subsequent diagrams.  
+![run-transaction-endpoint transact() method](docs/architecture/images/run-transaction-endpoint-transact.png)  
+The above diagram shows the sequence diagram of transact() method of the PluginLedgerConnectorPolkadot class. The caller to this function, which in reference to the above sequence diagram is API server, sends RunTransactionRequest object as an argument to the transact() method. Based on the type of Web3SigningCredentialType, corresponding responses are sent back to the caller.  
+![run-transaction-endpoint transactCactusKeychainRef() method](docs/architecture/images/run-transaction-endpoint-transact-cactuskeychainref.png)  
+The above diagram shows transactCactusKeychainReference() method being called by the transact() method of the PluginLedgerConnector class when the Web3SigningCredentialType is CACTUSKEYCHAINREF. This method inturn calls transactMnemonicString() which calls the signAndSend() method of the Polkadot library.  
+![runtransaction-endpoint transactMnemonicString() method](docs/architecture/images/run-transaction-endpoint-transact-mnemonicstring.png)  
+The above diagram shows transactMnemonicString() method being called by the transact() method of the PluginLedgerConnector class when the Web3SigningCredentialType is MNEMONICSTRING. This method then calls the signAndSend() method of the Polkadot library.  
+![run-transaction-endpoint transactSigned() method](docs/architecture/images/run-transaction-endpoint-transact-signed.png)  
+The above diagram shows transactSigned() method being called by the transact() method of the PluginLedgerConnector class when the Web3SigningCredentialType is NONE. This method calls the api.rpc.author.submitAndWatchExtrinsic() of the Polkadot library.  
+
+
+## 3. Containerization
+### 3.1. Building/running the container image locally
+
+In the Cactus project root say:
+
+```sh
+DOCKER_BUILDKIT=1 docker build -f ./packages/cactus-plugin-ledger-connector-polkadot/Dockerfile . -t cplcb
```

Comment:
@AnmolBansalDEV I just tried to run this and it failed with the log pasted below in my comment.

One thing I noticed that you are still installing the old yarn v1 but we've migrated to v3 in the meantime and that ships with NodeJS out of the box too, IIRC.
What I recommend is taking a look at the `Dockerfile` of `cmd-api-server` and taking a cue from over there (you can also use the newer versions of that as a base image)

```sh
$ DOCKER_BUILDKIT=1 docker build -f ./packages/cactus-plugin-ledger-connector-polkadot/Dockerfile . -t cplcb
[+] Building 1.5s (6/6) FINISHED                                                                                                                                                                                                                  
 => [internal] load build definition from Dockerfile                                                                                                                                                                                         0.0s
 => => transferring dockerfile: 309B                                                                                                                                                                                                         0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                            0.0s
 => => transferring context: 887B                                                                                                                                                                                                            0.0s
 => [internal] load metadata for ghcr.io/hyperledger/cactus-cmd-api-server:2022-08-05-7309f2a                                                                                                                                                0.5s
 => [1/3] FROM ghcr.io/hyperledger/cactus-cmd-api-server:2022-08-05-7309f2a@sha256:059d1cc7312ff4a9d10361003176b1be8080a4e0c992c4863c28caf27db60465                                                                                          0.0s
 => CACHED [2/3] RUN npm install -g yarn@1.22.17                                                                                                                                                                                             0.0s
 => ERROR [3/3] RUN yarn add @hyperledger/cactus-plugin-ledger-connector-polkadot@latest --production --ignore-engines                                                                                                                       0.9s
------                                                                                                                                                                                                                                            
 > [3/3] RUN yarn add @hyperledger/cactus-plugin-ledger-connector-polkadot@latest --production --ignore-engines:                                                                                                                                  
#5 0.452 yarn add v1.22.17                                                                                                                                                                                                                        
#5 0.462 warning package.json: No license field                                                                                                                                                                                                   
#5 0.516 warning No license field
#5 0.518 [1/4] Resolving packages...
#5 0.807 error An unexpected error occurred: "https://registry.yarnpkg.com/@hyperledger%2fcactus-plugin-ledger-connector-polkadot: Not found".
#5 0.807 info If you think this is a bug, please open a bug report with the information provided in "/usr/src/app/yarn-error.log".
#5 0.807 info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
------
executor failed running [/bin/bash -c yarn add @hyperledger/cactus-plugin-ledger-connector-polkadot@${NPM_PKG_VERSION} --production --ignore-engines]: exit code: 1
```

Reply:
updated the name of the file, `cmd-api-server` package version, also the yarn version.
https://github.com/hyperledger/cacti/blob/87ec994040fade0e7a4a690883c971baa9eb9702/packages/cactus-plugin-ledger-connector-polkadot/Dockerfile#L1-L7

I installed the node v20 in a Ubuntu container, but it didn't come with `yarn` pre-installed.

>#5 0.807 error An unexpected error occurred: "https://registry.yarnpkg.com/@hyperledger%2fcactus-plugin-ledger-connector-polkadot: Not found".
the log mentions that the `@hyperledger/cactus-plugin-ledger-connector-polkadot` package is not found, which is true as it is not published to NPM registry. Please publish the package.

Reply:
@petermetz I've updated the Dockerfile to remove some issues while building it. It's now building successfully with the published canary version `@hyperledger/cactus-plugin-ledger-connector-polkadot@2.0.0-lfx-connector.163` but it runs into an error during startup. I think it's due to the absence of `onPluginInit()` method cause the `cactus-cmd-api-server` expects it. With the release of a version with new changes, I think it should work fine. I've updated the README.md regarding running the container as well.

https://github.com/hyperledger/cacti/blob/f7e7a2a1cd9694327e4d323650bd3fad9b3b80dd/packages/cactus-plugin-ledger-connector-polkadot/Dockerfile#L1-L10

Reply:
@AnmolBansalDEV Gotcha, that makes sense. I've done a lot of publishing just now.
npm package: `@hyperledger/cactus-plugin-ledger-connector-polkadot@2.0.0-lfx-connector.164`
connector container: `ghcr.io/hyperledger/cactus-connector-polkadot:2024-01-16--2.0.0-lfx-connector.164--pr-2877`
ledger container: `ghcr.io/hyperledger/cactus-substrate-all-in-one:2024-01-16-pr-2877`

The ledger container image was built with the npm package referenced through the build args, e.g. it has the latest code.

Please re-test everything with these and let me know how it goes (e.g. pass it back for review if you need something from me or if you think everything is working now and it's ready for the final review)

Reply:
@petermetz 
The `connector container`  is giving an unauthorized error to me
logs:
![image](https://github.com/hyperledger/cacti/assets/86472377/4e64afcf-c230-4fe4-a603-2b66061f03b4)
while the `ledger container` is working fine
Please make the image docker image public for me to test it. Thanks!

Reply:
> @petermetz The `connector container` is giving an unauthorized error to me logs: ![image](https://private-user-images.githubusercontent.com/86472377/298798022-4e64afcf-c230-4fe4-a603-2b66061f03b4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5OTgxODMsIm5iZiI6MTcwNTk5Nzg4MywicGF0aCI6Ii84NjQ3MjM3Ny8yOTg3OTgwMjItNGU2NGFmY2YtYzIzMC00ZmU0LWE2MDMtMmI2NjA2MWYwM2I0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIzVDA4MTgwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiNzhmMGZhZTg4YWI2YTA4ZmQ2MDcyN2FlNDgwMjJlYmU0NTM3ODEzNjgxYTAwYWI5MTJlMjljOTJmM2U5MDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.yT0obxAhxneoRQMShnZXZexx_xB6LphvwTWm0jVrOxc) while the `ledger container` is working fine Please make the image docker image public for me to test it. Thanks!

@AnmolBansalDEV Oh no, I'm so sorry I forgot to mark it as public on the official repo. Please try again now, it should be visible!

Reply:
@petermetz I've updated the connector with some changes and now it's working. Can you please publish it and then review the PR?

Reply:
> @petermetz I've updated the connector with some changes and now it's working. Can you please publish it and then review the PR?

@AnmolBansalDEV Sure thing, but which one do you mean? The connector container image or the connector npm package?

Reply:
@petermetz I meant the npm package, but I think we'll still have to rebuilt the connector image with the new npm package as the input. So all in all you'll have to publish a new npm package version and then rebuilt the docker image of the connector and then publish it to GitHub docker registery.

Reply:
@AnmolBansalDEV Gotcha, I just did both, you c an find the container image here: `ghcr.io/hyperledger/cactus-connector-polkadot:2024-02-07--2.0.0-lfx-connector.208--pr-2877`

Reply:
Thank you @petermetz

Reply:
> Thank you @petermetz

@AnmolBansalDEV My pleasure and thank you again for the contribution!